minidump-common/src/format.rs

```diff
@@ -1634,3 +1635,213 @@ pub struct DSO_DEBUG_64 {
     /// The address of the "dynamic structure"
     pub dynamic: u64,
 }
+
+/// A variable-length UTF-8-encoded string carried within a minidump file.
+#[derive(Clone)]
+pub struct MINIDUMP_STRING {
+    /// The length of the #Buffer field in bytes, not including the `NUL` terminator.
+    ///
+    /// This field is interpreted as a byte count, not a count of Unicode code points.
+    pub length: u32,
+    /// The string, encoded in UTF-8, and terminated with a `NUL` byte.
+    pub buffer: Vec<u8>,
+}
+
+impl<'a> scroll::ctx::TryFromCtx<'a, Endian> for MINIDUMP_STRING {
+    type Error = scroll::Error;
+
+    fn try_from_ctx(src: &[u8], endian: Endian) -> Result<(Self, usize), Self::Error> {
+        let offset = &mut 0;
+        let length: u32 = src.gread_with(offset, endian)?;
+        let data: &[u8] = src.gread_with(offset, length as usize)?;
```

Comment:
This actually shows two separate issues; I'll fix them momentarily.

1) This is dead code. I implemented this and then noticed that the way we read minidumps currently we have helper functions in `minidump.rs`. I'd appreciate some input on what you would prefer at this point. 

In a follow-up, I would probably refactor the minidump module to deserialize into usable types directly. Right now, there is an almost-Rust implementation of the format, and then a wrapper. See [this](https://github.com/willglynn/pdb/blob/547e825dc08fd4fbb883917d3466b0fb4b3560e9/src/symbol/mod.rs#L343-L378) for an example of such such a pattern.

2) This implementation reads too little and the other implementation doesn't validate the NUL byte. Both need fixing.

Reply:
> In a follow-up, I would probably refactor the minidump module to deserialize into usable types directly. Right now, there is an almost-Rust implementation of the format, and then a wrapper

I'm conflicted about this. I agree that having two parallel structs is a bit of a pain, but I really like the fact that the structs in `format.rs` map directly to the data on disk for the most part. I don't think you get that same benefit in `pdb` because the PDB file format is quite a bit different. The one similarity is that both formats have data streams that are composed of primitive data types but also contain references to other data types that you need to read to really make use of them, like `MINIDUMP_MODULE_LIST` containing a list of `MINIDUMP_MODULE` objects but then needing to go read the name and CodeView record for each module separately. The APIs in this crate don't necessarily do a great job of leveraging the distinction, but I know that for certain operations on PDB files it's critical to not parse entire data streams up-front, but do so iteratively or using the hashtables in the file for performance.

I'm open to suggestions, especially since you're actively working on the codebase now and I'm not.

Reply:
You got a point here, which would be especially interesting once we implement a Minidump _writer_. In that case, it's definitely useful to maintain raw structs and then have an easy streamlined mapping to the higher-level types. We can definitely draw a lot of inspiration from `goblin` for this.

I'll follow up about this in a separate issue over the next couple of days to collect thoughts.

Reply:
My former Mozilla colleagues [have started work on a Rust crate for minidump writing](https://phabricator.services.mozilla.com/D98744). It doesn't look like it uses this crate directly currently, but maybe vendors parts of it? I've inquired about the reasoning for that.